src/Microsoft.ML.Transforms/GcnTransform.cs

```diff
@@ -225,477 +255,725 @@ public void Save(ModelSaveContext ctx)
-        /// based on centered values (i.e. after subtracting the mean). But since the centered
-        /// values mean is approximately zero, we can use variance of non-centered values.
+        /// The kind of unit norm vectors are rescaled to. This enumeration is serialized.
         /// </summary>
-        private static Float StdDev(Float[] values, int count, int length)
-        {
-            Contracts.Assert(0 <= count && count <= length);
-            if (count == 0)
-                return 0;
-            // We need a mean to compute variance.
-            Float tmpMean = CpuMathUtils.Sum(values.AsSpan(0, count)) / length;
-            Float sumSq = 0;
-            if (count != length && tmpMean != 0)
-            {
-                // Sparse representation.
-                Float meanSq = tmpMean * tmpMean;
-                sumSq = (length - count) * meanSq;
-            }
-            sumSq += CpuMathUtils.SumSq(tmpMean, values.AsSpan(0, count));
-            return MathUtils.Sqrt(sumSq / length);
+        public enum NormalizerKind : byte
+        {
+            L2Norm = 0,
+            StdDev = 1,
+            L1Norm = 2,
+            LInf = 3
+        }
+
+        internal static class Defaults
+        {
+            public const NormalizerKind NormKind = NormalizerKind.L2Norm;
+            public const bool LpSubMean = false;
+            public const bool GcnSubMean = true;
+            public const bool UseStdDev = false;
+            public const float Scale = 1;
         }
 
         /// <summary>
-        /// Compute Standard Deviation.
-        /// We have two overloads of StdDev instead of one with <see cref="Nullable{Float}"/> mean for perf reasons.
+        /// Create a <see cref="LpNormEstimatorBase"/> that takes multiple pairs of columns.
         /// </summary>
-        private static Float StdDev(Float[] values, int count, int length, Float mean)
+        public LpNormEstimatorBase(IHostEnvironment env, params LpNormalizingTransform.ColumnInfoBase[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(LpNormalizingEstimator)), new LpNormalizingTransform(env, columns))
         {
-            Contracts.Assert(0 <= count && count <= length);
-            if (count == 0)
-                return 0;
-            Float sumSq = 0;
-            if (count != length && mean != 0)
+
+        }
+
+        public override SchemaShape GetOutputSchema(SchemaShape inputSchema)
+        {
+            Host.CheckValue(inputSchema, nameof(inputSchema));
+            var result = inputSchema.Columns.ToDictionary(x => x.Name);
+            foreach (var colPair in Transformer.Columns)
             {
-                // Sparse representation.
-                Float meanSq = mean * mean;
-                sumSq = (length - count) * meanSq;
+                if (!inputSchema.TryFindColumn(colPair.Input, out var col))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colPair.Input);
```

Comment:
>ExceptSchemaMismatch [](start = 31, length = 20)

`ExceptSchemaMismatch`

Reply:
It's already ExceptSchemaMismatch, i don't get meaning of this comment.

---
In reply to: [232437178](https://github.com/dotnet/machinelearning/pull/1582#discussion_r232437178) [](ancestors = 232437178)